### PR TITLE
Add d2c function

### LINKED ISSUE #347: How do I convert from  discrete to continuous time
I think in Matlab there is a way to convert from discrete time model to continous time model

sysc = d2c(sysd)
sysc = d2c(sysd,method)

Description
example
sysc = d2c(sysd) converts a the discrete-time dynamic system model sysd to a continuous-time model using zero-order hold on the inputs.

example
sysc = d2c(sysd,method) specifies the conversion method.

How do I do it in ControlSystems.jl

### PR DESCRIPTION
**Pull request by baggepinnen**
This version of `d2c` is the inverse of `c2d` for ZoH and handles the case when `det(A-I) = 0`
closes #347 
this replaces #349 which does not handle this common case.

```julia
julia> sys = ss([0 1; 0 0], [0;1], [1 0], 0)
StateSpace{Continuous,Int64,Array{Int64,2}}
A = 
 0  1
 0  0
B = 
 0
 1
C = 
 1  0
D = 
 0

Continuous-time state-space model

julia> sysd = c2d(sys, 1)[1]
StateSpace{Discrete{Int64},Float64,Array{Float64,2}}
A = 
 1.0  1.0
 0.0  1.0
B = 
 0.5
 1.0
C = 
 1.0  0.0
D = 
 0.0

Sample Time: 1 (seconds)
Discrete-time state-space model


julia> d2c(sysd)
StateSpace{Continuous,Float64,Array{Float64,2}}
A = 
 0.0  1.0
 0.0  0.0
B = 
 -9.71445146547012e-17
  1.0
C = 
 1.0  0.0
D = 
 0.0

Continuous-time state-space model

julia> @test d2c(sysd) ≈ sys
Test Passed

julia> det(sysd.A - I)
0.0
```

### CODE REVIEW

**Comment by StevenSiew** [comment]
I made some minor changes

1.  import SpecialFunctions: factorial
2.  E = sum([ ((A_c * sys.timeevol.Ts)^k)/factorial(Float64(k + 1))  for k = 0:49])

Basically, express the factorial in Float64 instead of Int64, loop k from 0 to 49 instead of 0 to 19
and make sure we use the factorial function from SpecialFunctions

See issue https://github.com/JuliaControl/ControlSystems.jl/issues/347

```julia
import SpecialFunctions: factorial

"""`sysc = d2c(sys, method=:zoh)`
Convert the discrete system `sys` into a continous system using
the provided method. Currently only `:zoh` are provided.
Returns the discrete system `sysc`
"""
function d2c(sys::StateSpace,method::Symbol=:zoh)
    if !isdiscrete(sys)
        error("sys must be a discrete time system")
    end
    val_1 = one( eltype(sys.A) )
    if method == :zoh
        if det(sys.A - I) == 0
            # Use alternative method for :zoh because det(A - I) == 0")
            A_c = real.(  (val_1/sys.timeevol.Ts) .* log(sys.A)  )
            E = sum([ ((A_c * sys.timeevol.Ts)^k)/factorial(Float64(k + 1))  for k = 0:49])
            B_c = (val_1/sys.timeevol.Ts) .* inv(E) * sys.B
            C_c = sys.C
            D_c = sys.D
            return ss(A_c,B_c,C_c,D_c)
        end
        A_c = real.(  (val_1/sys.timeevol.Ts) .* log(sys.A)  )
        B_c = inv(sys.A - I) * A_c * sys.B
        C_c = sys.C
        D_c = sys.D
        return ss(A_c,B_c,C_c,D_c)
    elseif method == :foh || method == :tustin || method == :matched
        error("NotImplemented: Only `:zoh` is implemented so far")
    else
        error("Unsupported method: ", method)
    end
end

function d2c(G::TransferFunction,method::Symbol=:zoh)
    if !isdiscrete(G)
        error("TransferFunc must be a discrete time system")
    end
    return tf(  d2c(ss(G),method)  )
end
```

**Comment by baggepinnen** [comment]
See #355 which adds a much simpler version of `d2c` which also handles `det(A-I) = 0`.

**Comment by baggepinnen** [comment]
@StevenSiew  this PR is a new PR opened by me. The PR you opened is #349

**Comment by StevenSiew** [comment]
Okay I see. Your version of the source code is much better than mine. Thank you for your assistance. I still trying to figure out

```
        M = log([A  B;
            zeros(nu, nx) I])./sys.Ts
        Ac = M[1:nx, 1:nx]
        Bc = M[1:nx, nx+1:nx+nu]
```

**Comment by baggepinnen** [comment]
Test failure affects julia v1.0 which seem to have different numerical properties.

File: src/discrete.jl [hunk 1]
```diff
@@ -44,6 +44,29 @@ function c2d(sys::StateSpace, Ts::Real, method::Symbol=:zoh)
...
+    nx = nstates(sys)
+    if method == :zoh
+        M = log([A  B;
+            zeros(nu, nx) I])./sys.Ts
+        Ac = M[1:nx, 1:nx]
+        Bc = M[1:nx, nx+1:nx+nu]
+        Cc = C
+        Dc = D
+        if eltype(A) <: Real
+            Ac,Bc = real.((Ac, Bc))
```
Comment (mfalt) [inline src/discrete.jl:59]: What does this do? If real make real?

Comment (baggepinnen) [reply src/discrete.jl:59]: Yeah, log will always return complex matrices, but if A was real, then Ac should also be.

Comment (mfalt) [reply src/discrete.jl:59]: Ah, I missed that it didn't check `eltype(Ac) <: Real`

**Comment by mfalt** [comment]
My suggestion is that we add @test_broken with
`@static if VERSION < v"1.4"` until they release 1.0.6.
After that we could remove it and change compat to
julia 1.0.6+
julia 1.5
And possibly change it even more depending on what we decide in https://github.com/JuliaControl/ControlSystems.jl/pull/356

**Pull request merged**